src/rand.rs

```diff
@@ -0,0 +1,77 @@
+use num_bigint::{BigUint, RandBigInt};
+use rand::Rng;
+
+/// Iterator to generate a given amount of random numbers. For convenience of
+/// use with miller_rabin tests, you can also append a specified number at the
+/// end of the generated stream.
+pub struct Randoms<R> {
+    appended: Option<BigUint>,
+    lower_limit: BigUint,
+    upper_limit: BigUint,
+    amount: usize,
+    rng: R,
+}
+
+impl<R: Rng> Randoms<R> {
+    pub fn new(lower_limit: BigUint, upper_limit: BigUint, amount: usize, rng: R) -> Self {
+        Self {
+            appended: None,
+            lower_limit,
+            upper_limit,
+            amount,
+            rng,
+        }
+    }
+
+    /// Append the number at the end to appear as if it was generated. This
+    /// doesn't affect stream length. Only one number can be appended,
+    /// subsequent calls will replace the previously appended number.
+    pub fn with_appended(mut self, x: BigUint) -> Self {
+        self.appended = Some(x);
+        self
+    }
+
+    fn gen_biguint(&mut self) -> BigUint {
+        self.rng
+            .gen_biguint_range(&self.lower_limit, &self.upper_limit)
+    }
+}
+
+impl<R: Rng> Iterator for Randoms<R> {
+    type Item = BigUint;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.amount == 0 {
+            None
+        } else if self.amount == 1 {
+            let r = match core::mem::replace(&mut self.appended, None) {
```

Comment:
Thanks for the PR

This part seems a bit kludgy. I wonder if there’s a better way to do this.
I’ll think about it a little bit more and either recommend an alternative or do it myself after merging